src/NServiceBus.Core/Utils/Reflection/DelegateFactory.cs

```diff
@@ -61,13 +61,22 @@ static class DelegateFactory
                 var gen = method.GetILGenerator();
 
                 gen.Emit(OpCodes.Ldarg_0); // Load input to stack
-                gen.Emit(OpCodes.Castclass, sourceType); // Cast to source type
+
+                if (sourceType.IsValueType)
+                {
+                    gen.Emit(OpCodes.Unbox, sourceType);  //this is a struct so unbox
```

Comment:
I'd have to study the IL more, but it does seem a bit strange that we end up having an `Unbox` here and then an `Unbox_Any` on line 75 as well, almost as if we're unboxing twice?

Reply:
I'll test with a single unbox

Reply:
Looking at this I think the code is correct since one is the unbox for the source type and the other is for the actual value to set?